File: src/main/java/com/patrickanker/chatterbox/Chatterbox.java
package com.patrickanker.chatterbox;

import com.patrickanker.chatterbox.core.ChannelManager;
import org.bukkit.Bukkit;
import org.bukkit.plugin.Plugin;

import java.util.logging.Level;

public final class Chatterbox {

    private static final String LOGGING_ID = "Chatterbox";
    private static Chatterbox instance;

    private String executingPlugin;
    private PermissionsManager permissionsManager;
    private ChannelManager channelManager;

    public static Chatterbox chatterbox() {
        if (instance != null) {
            return instance;
        } else {
            instance = new Chatterbox();

            return instance;
        }
    }

    public static void start(Plugin plugin) {
        // Loading logic here
        chatterbox(); // Create the instance

        instance.executingPlugin = plugin.getName();
    }

    public static void stop() {
        // Stopping logic here
    }

    public static Plugin executingPlugin() {
        return Bukkit.getServer().getPluginManager().getPlugin(chatterbox().executingPlugin);
    }

    public static PermissionsManager permissionsManager() {
        return chatterbox().permissionsManager;
    }

    public static ChannelManager channelManager() {
        return chatterbox().channelManager;
    }

    public static void logInfo(String message) {
        Bukkit.getLogger().log(Level.INFO, "[{0}] {1}", new Object[] {LOGGING_ID, message});
    }

    public static void logWarn(String message) {
        Bukkit.getLogger().log(Level.WARNING, "[{0}] {1}", new Object[] {LOGGING_ID, message});
    }

    public static void logSevere(String message) {
        Bukkit.getLogger().log(Level.SEVERE, "[{0}] {1}", new Object[] {LOGGING_ID, message});
    }
}

File: src/main/kotlin/chatterbox/Chatterbox.kt
package chatterbox

import chatterbox.api.Messenger
import chatterbox.util.commands.DynamicCommandRegistry
import org.bukkit.Bukkit
import org.bukkit.plugin.Plugin
import java.util.logging.Level

const val LOGGING_ID: String = "Chatterbox"

object Chatterbox {
    private val knownMessengers: MutableSet<Messenger> = mutableSetOf()
    private val commandRegistry by lazy {
        DynamicCommandRegistry(executingPlugin)
    }

    private var executingPlugin: String = ""

    fun start(plugin: Plugin) {
        executingPlugin = plugin.name

        // Configuration loading here

        // Command registration here
    }

    fun hasMessenger(id: String): Boolean = knownMessengers.firstOrNull { it.id == id } != null

    fun findMessenger(id: String): Messenger? = knownMessengers.firstOrNull { it.id == id }

    fun addMessenger(messenger: Messenger) {
        if (!hasMessenger(messenger.id)) {
            knownMessengers.add(messenger)
        }
    }

    fun removeMessenger(id: String) {
        if (hasMessenger(id)) {
            knownMessengers.remove(findMessenger(id)!!) // !! Check is valid from hasMessenger call
        }
    }
}

fun logInfo(message: String)   = Bukkit.getLogger().log(Level.INFO, "[$LOGGING_ID] $message")
fun logWarn(message: String)   = Bukkit.getLogger().log(Level.WARNING, "[$LOGGING_ID] $message")
fun logSevere(message: String) = Bukkit.getLogger().log(Level.SEVERE, "[$LOGGING_ID] $message")
